modules/module-postgres-storage/src/storage/PostgresCompactor.ts

```diff
@@ -0,0 +1,366 @@
+   */
+  opsSincePut: number;
+}
+
+/**
+ * Additional options, primarily for testing.
+ */
+export interface PostgresCompactOptions extends storage.CompactOptions {
+  /** Minimum of 2 */
+  clearBatchLimit?: number;
+  /** Minimum of 1 */
+  moveBatchLimit?: number;
+  /** Minimum of 1 */
+  moveBatchQueryLimit?: number;
+}
+
+const DEFAULT_CLEAR_BATCH_LIMIT = 5000;
+const DEFAULT_MOVE_BATCH_LIMIT = 2000;
+const DEFAULT_MOVE_BATCH_QUERY_LIMIT = 10_000;
+
+/** This default is primarily for tests. */
+const DEFAULT_MEMORY_LIMIT_MB = 64;
+
+export class PostgresCompactor {
+  private updates: pgwire.Statement[] = [];
+
+  private idLimitBytes: number;
+  private moveBatchLimit: number;
+  private moveBatchQueryLimit: number;
+  private clearBatchLimit: number;
+  private maxOpId: bigint | undefined;
+  private buckets: string[] | undefined;
+
+  constructor(
+    private db: lib_postgres.DatabaseClient,
+    private group_id: number,
+    options?: PostgresCompactOptions
+  ) {
+    this.idLimitBytes = (options?.memoryLimitMB ?? DEFAULT_MEMORY_LIMIT_MB) * 1024 * 1024;
+    this.moveBatchLimit = options?.moveBatchLimit ?? DEFAULT_MOVE_BATCH_LIMIT;
+    this.moveBatchQueryLimit = options?.moveBatchQueryLimit ?? DEFAULT_MOVE_BATCH_QUERY_LIMIT;
+    this.clearBatchLimit = options?.clearBatchLimit ?? DEFAULT_CLEAR_BATCH_LIMIT;
+    this.maxOpId = options?.maxOpId;
+    this.buckets = options?.compactBuckets;
+  }
+
+  /**
+   * Compact buckets by converting operations into MOVE and/or CLEAR operations.
+   *
+   * See /docs/compacting-operations.md for details.
+   */
+  async compact() {
+    if (this.buckets) {
+      for (let bucket of this.buckets) {
+        // We can make this more efficient later on by iterating
+        // through the buckets in a single query.
+        // That makes batching more tricky, so we leave for later.
+        await this.compactInternal(bucket);
+      }
+    } else {
+      await this.compactInternal(undefined);
+    }
+  }
+
+  async compactInternal(bucket: string | undefined) {
+    const idLimitBytes = this.idLimitBytes;
+
+    let currentState: CurrentBucketState | null = null;
+
+    let bucketLower: string | null = null;
+    let bucketUpper: string | null = null;
+
+    if (bucket?.includes('[')) {
+      // Exact bucket name
+      bucketLower = bucket;
+      bucketUpper = bucket;
+    } else if (bucket) {
+      // Bucket definition name
+      bucketLower = `${bucket}[`;
+      bucketUpper = `${bucket}[\uFFFF`;
+    }
+
+    let upperOpIdLimit = BIGINT_MAX;
+
+    while (true) {
+      const batch = await this.db.sql`
+        SELECT
+          op,
+          op_id,
+          source_table,
+          table_name,
+          row_id,
+          source_key,
+          bucket_name
+        FROM
+          bucket_data
+        WHERE
+          group_id = ${{ type: 'int4', value: this.group_id }}
+          AND bucket_name LIKE COALESCE(${{ type: 'varchar', value: bucketLower }}, '%')
```

Comment:
Not sure if this will efficiently use the primary key index for filtering bucket names if specified. Not important for intial release.